File: packages/proofs/src/asp_proofs/project_topology_frontier.py
# SPDX-FileCopyrightText: 2026 tao3k team and Contributors
#
# SPDX-License-Identifier: Apache-2.0 AND LGPL-2.1-or-later

"""Expected-relation and frontier truth table for Project Topology V1."""

from __future__ import annotations

from ._topology_admission import require_topology


def _key(item: dict) -> tuple[str, str, str, str, int]:
    return (
        item["anchor"],
        item["target"],
        item["relation"],
        item["targetKind"],
        item["depth"],
    )


def _coverage(
    reference: str | None,
    relation: str,
    target_kind: str,
    scope: str,
    certificates: dict[str, dict],
) -> str:
    require_topology(reference is not None, "topology-frontier-coverage-missing")
    certificate = certificates.get(reference)
    require_topology(
        certificate is not None
        and certificate["relation"] == relation
        and certificate["targetKind"] == target_kind
        and certificate["scope"] == scope,
        "topology-frontier-coverage-mismatch",
    )
    return reference


def validate_relation_frontiers(
    library: dict, nodes: dict[str, dict], edges: dict[str, dict]
) -> None:
    """Require frontiers to equal exactly unresolved source-owned obligations."""

    positive = {
        (edge["from"], edge["to"], edge["relation"])
        for edge in edges.values()
        if edge["modality"] in {"parser-direct", "declared", "derived"}
    }
    certificates: dict[str, dict] = {}
    for certificate in library["coverageCertificates"]:
        certificate_id = certificate["id"]
        require_topology(
            certificate_id not in certificates, "topology-coverage-duplicate"
        )
        certificates[certificate_id] = certificate

    frontiers: dict[tuple[str, str, str, str, int], dict] = {}
    for frontier in library["frontiers"]:
        key = _key(frontier)
        require_topology(key not in frontiers, "topology-frontier-duplicate")
        frontiers[key] = frontier

    expectation_ids: set[str] = set()
    expectation_keys: set[tuple[str, str, str, str, int]] = set()
    unresolved: set[tuple[str, str, str, str, int]] = set()
    referenced_certificates: set[str] = set()
    for expected in library["expectedRelations"]:
        expected_id = expected["id"]
        require_topology(
            expected_id not in expectation_ids,
            "topology-expected-relation-duplicate",
        )
        expectation_ids.add(expected_id)
        key = _key(expected)
        require_topology(
            key not in expectation_keys, "topology-expected-relation-duplicate"
        )
        expectation_keys.add(key)
        anchor, target, relation, target_kind, _depth = key
        require_topology(
            anchor in nodes and nodes.get(target, {}).get("kind") == target_kind,
            "topology-expected-relation-endpoint-mismatch",
        )
        frontier = frontiers.get(key)
        if (anchor, target, relation) in positive:
            require_topology(frontier is None, "topology-positive-frontier-conflict")
            continue
        unresolved.add(key)
        require_topology(frontier is not None, "topology-frontier-missing")
        state = frontier["state"]
        reason = frontier["reason"]
        reference = frontier.get("coverageRef")
        coverage = expected["coverage"]
        if coverage == "none":
            require_topology(
                (state, reason, reference)
                == ("unknown", "binding-not-established", None),
                "topology-frontier-classification-mismatch",
            )
        elif coverage == "partial":
            require_topology(
                (state, reason) == ("unknown", "coverage-open"),
                "topology-frontier-classification-mismatch",
            )
            referenced_certificates.add(
                _coverage(reference, relation, target_kind, "partial", certificates)
            )
        else:
            require_topology(
                (state, reason)
                == ("certified-missing", "complete-coverage-no-witness"),
                "topology-frontier-classification-mismatch",
            )
            referenced_certificates.add(
                _coverage(reference, relation, target_kind, "complete", certificates)
            )

    require_topology(
        set(frontiers) == unresolved, "topology-frontier-expectation-mismatch"
    )
    require_topology(
        referenced_certificates == set(certificates),
        "topology-frontier-coverage-extraneous",
    )

File: crates/agent-semantic-topology/tests/unit/project_topology_frontier.rs
// SPDX-FileCopyrightText: 2026 tao3k team and Contributors
//
// SPDX-License-Identifier: Apache-2.0 AND LGPL-2.1-or-later

use std::collections::BTreeMap;

use agent_semantic_topology::{ProjectTopologyLibrary, ProjectTopologyManifest};
use serde_json::{Value, json};

fn packet() -> Value {
    serde_json::from_str(include_str!(
        "../../../../schemas/fixtures/project-topology-library/valid-polyglot.v1.json"
    ))
    .expect("valid Project Topology fixture")
}

fn admit(packet: Value) -> Result<ProjectTopologyLibrary, String> {
    let manifest = ProjectTopologyManifest::parse_org(include_str!(
        "../../../../org/templates/project.workspace-manifest.v1.org"
    ))
    .expect("Project Topology manifest");
    let receipt = packet["fromScratchRebuildReceipt"].clone();
    let receipts = BTreeMap::from([("topology-rebuild-1".to_owned(), receipt)]);
    ProjectTopologyLibrary::admit_with_receipts(packet, manifest.project_workspace(), &receipts)
        .map_err(|error| error.reason_kind().to_owned())
}

fn unresolved(packet: &mut Value, coverage: &str) {
    packet["expectedRelations"] = json!([{
        "id": "expected-config",
        "anchor": "registry",
        "target": "refresh",
        "relation": "READS_CONFIG",
        "targetKind": "Method",
        "depth": 1,
        "coverage": coverage
    }]);
    let (state, reason) = match coverage {
        "none" => ("unknown", "binding-not-established"),
        "partial" => ("unknown", "coverage-open"),
        "complete" => ("certified-missing", "complete-coverage-no-witness"),
        _ => unreachable!("test coverage is known"),
    };
    packet["frontiers"] = json!([{
        "anchor": "registry",
        "target": "refresh",
        "relation": "READS_CONFIG",
        "targetKind": "Method",
        "depth": 1,
        "state": state,
        "reason": reason
    }]);
    if coverage != "none" {
        packet["frontiers"][0]["coverageRef"] = json!("coverage-config");
        packet["coverageCertificates"] = json!([{
            "id": "coverage-config",
            "relation": "READS_CONFIG",
            "targetKind": "Method",
            "scope": coverage,
            "digest": "blake3-256:9999999999999999999999999999999999999999999999999999999999999999"
        }]);
    }
}

#[test]
fn unresolved_relation_truth_table_is_admitted() {
    for coverage in ["none", "partial", "complete"] {
        let mut packet = packet();
        unresolved(&mut packet, coverage);
        admit(packet).expect("truth-table frontier should be admitted");
    }
}

#[test]
fn frontier_reason_must_match_coverage_truth_table() {
    let mut packet = packet();
    unresolved(&mut packet, "partial");
    packet["frontiers"][0]["reason"] = json!("binding-not-established");
    assert_eq!(
        admit(packet).expect_err("wrong reason must fail"),
        "topology-frontier-classification-mismatch"
    );
}

#[test]
fn proposed_edge_does_not_close_source_owned_expectation() {
    let mut packet = packet();
    unresolved(&mut packet, "none");
    let semantic_digest = packet["identities"]["semanticTopologyDigest"].clone();
    packet["edges"].as_array_mut().expect("edges").push(json!({
        "id": "proposed-config",
        "segmentId": null,
        "from": "registry",
        "to": "refresh",
        "relation": "READS_CONFIG",
        "modality": "proposed",
        "bindingDigest": semantic_digest,
        "witnesses": ["model-proposal-1"]
    }));
    admit(packet).expect("proposal leaves the source-owned frontier open");
}

#[test]
fn coverage_certificate_cannot_float_free_of_frontier() {
    let mut packet = packet();
    packet["coverageCertificates"] = json!([{
        "id": "coverage-unused",
        "relation": "READS_CONFIG",
        "targetKind": "Method",
        "scope": "partial",
        "digest": "blake3-256:8888888888888888888888888888888888888888888888888888888888888888"
    }]);
    assert_eq!(
        admit(packet).expect_err("unused coverage must fail"),
        "topology-frontier-coverage-extraneous"
    );
}
